Mathlib/CategoryTheory/Topos/Exponentials.lean

```diff
@@ -0,0 +1,480 @@
+      rw [prod.map_snd, prod.map_snd, Hom_comm]
+    have h_terminal :
+    (prod.map (𝟙 A) (terminal.from (Hom A B)) ≫ prod.fst) ≫ terminal.from A = terminal.from _ :=
+      Unique.eq_default _
+    rw [assoc, comm_middle, ←assoc, comm_left, assoc, Predicate.NameDef]
+    dsimp [Predicate.true_]
+    rw [←assoc, ←assoc, h_terminal]
+
+/-- The evaluation map eval : A ⨯ B^A ⟶ B. -/
+def eval (A B : C) : A ⨯ (Hom A B) ⟶ B :=
+  ClassifierCone_into (comm' := evalDef_comm A B)
+
+/-- This states the commutativity of the square relating
+`eval A B` to `singleton B` and `Hom_toGraph A B`, which
+arises from its definition.
+-/
+lemma evalCondition (A B : C) :
+    eval A B ≫ singleton B
+    = prod.map (𝟙 _) (Hom_toGraph A B) ≫ ((prod.associator _ _ _).inv ≫ in_ (B ⨯ A))^ :=
+  ClassifierCone_into_comm _ _ _
+
+/-- A map `f_exp : X ⟶ HomAB` exponentiates `f : A ⨯ X ⟶ B` with respect to
+an "evaluation map" `e : A ⨯ HomAB ⟶ B` if the following diagram commutes:
+```
+
+        A ⨯ X ---f----> B
+          |             ^
+          |            /
+          |           /
+    (𝟙 A) ⨯ f_exp    /
+          |         /
+          |        e
+          |       /
+          |      /
+          |     /
+          |    /
+          v   /
+        A ⨯ HomAB
+```
+-/
+abbrev Exponentiates {A B X HomAB : C}  (e : A ⨯ HomAB ⟶ B) (f : A ⨯ X ⟶ B) (f_exp : X ⟶ HomAB) :=
+  (prod.map (𝟙 _) f_exp) ≫ e = f
+
+/-- An object `HomAB` and a map `e : A ⨯ HomAB ⟶ B` form an exponential object for `A B : C`
+if, for any map `f : A ⨯ X ⟶ B`, there is a unique map `f_exp : A ⟶ PB` such that
+the following diagram commutes:
+```
+
+        A ⨯ X ---f----> B
+          |             ^
+          |            /
+          |           /
+    (𝟙 A) ⨯ f_exp    /
+          |         /
+          |        e
+          |       /
+          |      /
+          |     /
+          |    /
+          v   /
+        A ⨯ HomAB
+```
+-/
+class IsExponentialObject {A B HomAB : C} (e : A ⨯ HomAB ⟶ B) where
+  /-- The map that sends morphisms `A ⨯ X ⟶ B` to morphisms `X ⟶ Hom A B`. -/
+  exp : ∀ {X} (_ : A ⨯ X ⟶ B), X ⟶ HomAB
+  /-- `exp f` satisfies commutativity of the above diagram. -/
+  exponentiates : ∀ {X} (f : A ⨯ X ⟶ B), Exponentiates e f (exp f)
+  /-- `exp f` is the only map that satisfies the above commutativity condition. -/
+  unique' : ∀ {X} {f : A ⨯ X ⟶ B} {exp' : X ⟶ HomAB}, Exponentiates e f exp' → exp f = exp'
+
+/-- What it means for a pair `A B : C` to have an exponential object.
+See `IsExponentialObject`.
+-/
+class HasExponentialObject (A B : C) where
+  /-- The "internal hom" object. -/
+  HomAB : C
+  /-- The evaluation map. -/
+  e : A ⨯ HomAB ⟶ B
+  /-- `HomAB` and `e` form an exponential object for `A B : C`. -/
+  is_exp : IsExponentialObject e
+
+variable (C)
+
+/-- A category has exponential objects if each pair of objects therein has an exponential object. -/
+class HasExponentialObjects where
+  has_exponential_object (A B : C) : HasExponentialObject A B
```

Comment:
See `Exponentiable` and `CartesianClosed` in `CategoryTheory.Closed.Cartesian`. See also `CategoryTheory.Adjunction.PartialAdjoint` if you absolutely need `HasExponentialObject` instead of `Exponentiable`.